Make THPVariable_Unpack work on THPVariable too

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56150 Make it illegal to directly construct _TensorBase
* #56147 Give Tensor a trivial (for now) metaclass _TensorMeta
* #56016 Use THPVariable_Unpack in python_nccl
* #55799 Replace all direct cdata access with THPVariable_Unpack
* **#55798 Make THPVariable_Unpack work on THPVariable too**
* #55797 Modify a few APIs to take/return const Tensor& instead of Tensor&

I'm going to change how cdata is implemented internally, so I want to
make all callsites call through THPVariable_Unpack even if they
actually have a THPVariable in hand

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712131](https://our.internmc.facebook.com/intern/diff/D27712131)

ezyang committed:
Make THPVariable_Unpack work on THPVariable too

I'm going to change how cdata is implemented internally, so I want to
make all callsites call through THPVariable_Unpack even if they
actually have a THPVariable in hand

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh commented on torch/csrc/autograd/python_variable.h:
```diff
@@ -46,7 +46,10 @@ inline bool THPVariable_Check(PyObject *obj)
-inline torch::autograd::Variable& THPVariable_Unpack(PyObject* obj) {
-  auto var = (THPVariable*)obj;
+inline const at::Tensor& THPVariable_Unpack(THPVariable* var) {
   return var->cdata;
```
It took me a second to remember that we have a `using Variable = at::Tensor` somewhere :p I thought we were doing some implicit cast at first, since `cdata` is of type Variable but we're returning a Tensor. This probably isn't the time, but do we eventually want to kill any references to Variable and just use Tensor everywhere?

ezyang replied:
Yeah, we should do that... some day... LOL

ezyang committed:
Update on "Make THPVariable_Unpack work on THPVariable too"


I'm going to change how cdata is implemented internally, so I want to
make all callsites call through THPVariable_Unpack even if they
actually have a THPVariable in hand

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712131](https://our.internmc.facebook.com/intern/diff/D27712131)

[ghstack-poisoned]

ezyang committed:
Update on "Make THPVariable_Unpack work on THPVariable too"


I'm going to change how cdata is implemented internally, so I want to
make all callsites call through THPVariable_Unpack even if they
actually have a THPVariable in hand

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712131](https://our.internmc.facebook.com/intern/diff/D27712131)

[ghstack-poisoned]

ezyang committed:
Update on "Make THPVariable_Unpack work on THPVariable too"


I'm going to change how cdata is implemented internally, so I want to
make all callsites call through THPVariable_Unpack even if they
actually have a THPVariable in hand

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712131](https://our.internmc.facebook.com/intern/diff/D27712131)

[ghstack-poisoned]
